src/ucp/api/cuda/ucp.cuh

```diff
@@ -0,0 +1,210 @@
+     * transfers.
+     */
+    uct_ep_t         **uct_ep;
+
+    /**
+     * Number of UCT exported endpoints found in @ref uct_ep arrays.
+     */
+    unsigned         num_uct_eps;
+
+    /**
+     * Number of entries in the descriptor list array @ref elems.
+     */
+    size_t           dlist_length;
+
+    /**
+     * Array of descriptor list containing memory pairs to be used by GPU
+     * device functions for memory transfers.
+     */
+    ucp_dlist_elem_t elems[];
+} ucp_dlist_handle_t;
+
+
+typedef ucp_dlist_handle_t *ucp_dlist_handle_h;
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Posts one memory put operation.
+ *
+ * This GPU routine posts one put operation using descriptor list handle.
+ * The @ref dlist_index is used to point at the dlist entry to be used for the
+ * memory transfer. The addresses and length must be valid for the used dlist
+ * entry.
+ *
+ * This routine can be called repeatedly with the same handle and different
+ * addresses and length. The flags parameter can be used to modify the behavior
+ * of the routine.
+ *
+ * @param [in]  handle      Exported descriptor list handle to use for transfer.
+ * @param [in]  addr        Local virtual address to send data from.
+ * @param [in]  remote_addr Remote virtual address to send data to.
+ * @param [in]  length      Length in bytes of the data to send.
+ * @param [in]  dlist_index Index in descriptor list pointing to the memory
+ *                          registration keys to use for the transfer.
+ * @param [in]  flags       Bitfield usable to modify the function behavior.
+ * @param [out] status      GPU status returned by the call.
+ */
+__device__ void
+ucp_gpu_post_single(ucp_dlist_handle_h handle,
+                    void *addr, uint64_t remote_addr,
+                    size_t length, int dlist_index, uint64_t flags,
+                    ucs_gpu_status_t *status);
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Posts one memory atomic increment operation.
+ *
+ * This GPU routine posts one atomic increment operation using descriptor list
+ * handle. The @ref dlist_index is used to point at the dlist entry to be used
+ * for the atomic operation. The remote address must be valid for the used dlist
+ * entry.
+ *
+ * This routine can be called repeatedly with the same handle and different
+ * address. The flags parameter can be used to modify the behavior of the
+ * routine.
+ *
+ * @param [in]  handle      Exported descriptor list handle to use for transfer.
+ * @param [in]  value       Value used to increment the remote address.
+ * @param [in]  remote_addr Remote virtual address to perform the increment to.
+ * @param [in]  dlist_index Index in descriptor list pointing to the memory
+ *                          remote key to use for the atomic operation.
+ * @param [in]  flags       Bitfield usable to modify the function behavior.
+ * @param [out] status      GPU status returned by the call.
+ */
+__device__ void
```

Comment:
Just a thought, and I am not sure whether its really practical or not. Shall we introduce an abstraction for the keywords such as `__device__` that would allow for GPU programming model agnostic API specification?  My line of thinking is that the abstract `UCP_DEVICE_FUNC` keyword or similar could be mapped to `__device__` if during configure time UCX enables CUDA (or ROCm for that matter), but might map to something else for SYCL or similar.